### PR TITLE
Remove inventory units

### PR DESCRIPTION
**Pull request by mbj**
Depends on:

* [x] #332

### CODE REVIEW

File: core/app/models/spree/stock/content_item.rb [hunk 1]
```diff
@@ -1,34 +1,22 @@
 module Spree
   module Stock
     class ContentItem
-      include Adamantium::Flat, Concord::Public.new(:inventory_unit)
+      include Adamantium::Flat, Concord::Public.new(:line_item)
 
       def variant
-        inventory_unit.variant
+        line_item.variant
```
Comment (dkubb) [inline core/app/models/spree/stock/content_item.rb]: You can replace this with `delegate :variant, to: :line_item`.

File: core/app/models/spree/stock/content_item.rb [hunk 2]
```diff
@@ -1,34 +1,22 @@
...
-        inventory_unit.line_item
-      end
-
-      def price
-        variant.price
-      end
-
-      def amount
-        price * quantity
+        line_item.line_item
```
Comment (dkubb) [inline core/app/models/spree/stock/content_item.rb]: Did you actually mean this? This is a recursive call.

File: core/app/models/spree/stock/content_item.rb (hunk 1 above)
Comment (mbj) [inline core/app/models/spree/stock/content_item.rb]: @dkubb Sorry forgot the `in progress` tag.

as this happens to often: Lets change branches to be not reviewable unless a specific tag exists?

File: core/app/models/spree/stock/content_item.rb [hunk 3]
```diff
@@ -1,34 +1,22 @@
...
       end
 
       def quantity
-        # Since inventory units don't have a quantity,
-        # make this always 1 for now, leaving ourselves
-        # open to a different possibility in the future,
-        # but this massively simplifies things for now
-        1
+        line_item.quantity
```
Comment (dkubb) [inline core/app/models/spree/stock/content_item.rb]: This can also be replaced with a delegate call.

File: core/app/models/spree/stock/availability_validator.rb [hunk 4]
```diff
@@ -2,14 +2,7 @@ module Spree
...
       def validate(line_item)
-        unit_count = line_item.inventory_units.size
-        return if unit_count >= line_item.quantity
-        quantity = line_item.quantity - unit_count
-        return if quantity.zero?
-
-        quantifier = Stock::Quantifier.new(line_item.variant)
-
-        return if quantifier.can_supply?(quantity)
+        return if line_item.quantity.zero? || Stock::Quantifier.new(line_item.variant).can_supply?(line_item.quantity)
```
Comment (dkubb) [inline core/app/models/spree/stock/availability_validator.rb]: Since `line_item.quantity` is referenced twice here I would suggest copying it into a local variable and referencing it in both places.

File: core/spec/models/spree/order_spec.rb [hunk 5]
```diff
@@ -757,11 +751,21 @@ def assert_expected_order_state
...
   describe "#create_proposed_shipments" do
-    it "assigns the coordinator returned shipments to its shipments" do
-      shipment = build(:shipment)
-      allow_any_instance_of(Spree::Stock::Coordinator).to receive(:shipments).and_return([shipment])
+    subject(:order) { create(:order) }
+
+    it 'creates shipments with rates from estimator' do
+      stock_location = create(:stock_location)
+      shipping_method = create(:shipping_method)
+      rate = shipping_method.shipping_rates.new(cost: 10)
```
Comment (dkubb) [inline core/spec/models/spree/order_spec.rb]: I would suggest aligning the equal signs here.

File: core/spec/models/spree/order_spec.rb [hunk 6]
```diff
@@ -757,11 +751,21 @@ def assert_expected_order_state
...
+      expect(Spree::Stock::Estimator).to receive(:call).and_return([rate])
+
       subject.create_proposed_shipments
-      expect(subject.shipments).to eq [shipment]
+
+      shipment = subject.shipments.first!
+      expect(shipment.shipping_method).to be(shipping_method)
+      expect(shipment.shipping_rates.first!).to be(rate)
+      expect(shipment.shipping_rates.size).to be(1)
```
Comment (dkubb) [inline core/spec/models/spree/order_spec.rb]: I would suggest checking the size before checking the value of the list.

Comment (mbj) [inline core/spec/models/spree/order_spec.rb]: Instead I used:

```
expect(shipment.shipping_rates.to_a).to eql([rate])
```

that effectively does both checks.

**Comment by mbj** [comment]
@dkubb This is ready for review.

**Comment by mbj** [comment]
> @dkubb This is ready for review.

Actually not, I still have to cleanup some stuff, sorry for the noise.

**Comment by mbj** [comment]
This was validated via:

```
RAILS_ENV=test bundle exec mutant -r ./spec/dummy/config/environment.rb -j 1 --use rspec -- 'Spree::Stock::Estimator*'
```

**Comment by mbj** [comment]
@dkubb This is ready for code only review. But not for merge. I do want to make sure the followup passes against this branch before we consider merging it.

**Comment by dkubb** [comment]
@mbj this is a big diff to review. Are there any other PRs we can extract from this to pare it down, or is this as small as it can possibly be?

File: backend/config/reek.yml [hunk 7]
```diff
@@ -13,7 +13,6 @@ ClassVariable:
 ControlParameter:
   enabled: true
   exclude:
-  - Spree::Admin::InventorySettingsHelper#show_not # true_or_false
```
Comment (dkubb) [inline backend/config/reek.yml]: Could this method removal be in it's own commit?

File: core/app/models/spree/order.rb [hunk 8]
```diff
@@ -530,10 +535,6 @@ def restart_checkout_flow
       self.next! if self.line_items.size > 0
     end
 
-    def refresh_shipment_rates
```
Comment (dkubb) [inline core/app/models/spree/order.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66472983

File: core/app/models/spree/app_configuration.rb [hunk 9]
```diff
@@ -31,7 +31,6 @@ class AppConfiguration < Preferences::Configuration
     preference :always_include_confirm_step, :boolean, default: false # Ensures confirmation step is always in checkout_progress bar, but does not force a confirm step if your payment methods do not support it.
     preference :always_put_site_name_in_title, :boolean, default: true
     preference :auto_capture, :boolean, default: false # automatically capture the credit card (as opposed to just authorize and capture later)
-    preference :binary_inventory_cache, :boolean, default: false # only invalidate product cache when a stock item changes whether it is in_stock
```
Comment (dkubb) [inline core/app/models/spree/app_configuration.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66472983

File: core/app/models/spree/shipment.rb [hunk 10]
```diff
@@ -112,12 +93,6 @@ def shipping_method
       selected_shipping_rate.try(:shipping_method) || shipping_rates.first.try(:shipping_method)
     end
 
-    def to_package
```
Comment (dkubb) [inline core/app/models/spree/shipment.rb]: Isn't this required by spree_active_shipping? Would it mean that this PR needs to be paired with an active shipping change?

File: core/app/models/spree/stock/availability_validator.rb [hunk 11]
```diff
@@ -2,14 +2,8 @@ module Spree
...
-        unit_count = line_item.inventory_units.size
-        return if unit_count >= line_item.quantity
-        quantity = line_item.quantity - unit_count
-        return if quantity.zero?
-
-        quantifier = Stock::Quantifier.new(line_item.variant)
-
-        return if quantifier.can_supply?(quantity)
+        quantity = line_item.quantity
+        return if quantity.zero? || Stock::Quantifier.new(line_item.variant).can_supply?(quantity)
```
Comment (dkubb) [inline core/app/models/spree/stock/availability_validator.rb]: If we change `spree_stock_items.count_on_hand` to have a CHECK constraint that asserts it is greater than or equal to 0, then we know `#can_supply?(0)` would always return `true`, so this extra `quantity.zero?` conditional isn't required.

The question is, should we defer adding a constraint we identified until a later PR?

IMHO we should always add constraints as we identify them, even if it means putting it in a PR that is merged ahead of this one. As soon as we find some invariant we know should hold for all records, we should add it to tighten up the system, and then revisit existing PRs and refactor them to be simpler.

File: core/spec/models/spree/order_spec.rb [hunk 12]
```diff
@@ -707,11 +701,21 @@ def assert_expected_order_state
...
-      shipment = build(:shipment)
-      allow_any_instance_of(Spree::Stock::Coordinator).to receive(:shipments).and_return([shipment])
+    subject(:order) { create(:order) }
+
+    it 'creates shipments with rates from estimator' do
+      stock_location  = create(:stock_location)
+      shipping_method = create(:shipping_method)
+      rate            = shipping_method.shipping_rates.new(cost: 10)
+
+      expect(Spree::Stock::Estimator).to receive(:call).and_return([rate])
```
Comment (dkubb) [inline core/spec/models/spree/order_spec.rb]: You could use `#receive_messages` here.

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 13]
```diff
@@ -1,106 +1,181 @@
...
 
-            allow(subject).to receive(:shipping_methods).and_return(shipping_methods)
+        before do
+          expect(calculator_a).to receive(:compute)
+            .with(package)
+            .and_return(cost)
 
-            expect(subject.shipping_rates(package).sort_by(&:cost).map(&:selected)).to eq [true, false]
+          expect(shipping_method_a).to receive(:shipping_rates)
+            .and_return(relation)
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66475113

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 14]
```diff
@@ -1,106 +1,181 @@
...
+        expect(calculator_a).to receive_messages(available?: true)
+        expect(calculator_b).to receive_messages(available?: true)
+
+        expect(calculator_a).to receive_messages(compute: shipping_rate_a.cost)
+        expect(calculator_b).to receive_messages(compute: shipping_rate_b.cost)
+
+        relation = double(:relation)
+
+        expect(shipping_method_a).to receive(:shipping_rates)
+          .and_return(relation)
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66475113

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 15]
```diff
@@ -1,106 +1,181 @@
...
+
+        expect(relation).to receive(:new)
+          .with(cost: shipping_rate_a.cost)
+          .and_return(shipping_rate_a)
+
+        expect(shipping_method_b).to receive(:shipping_rates)
+          .and_return(relation)
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66475113

File: core/app/models/spree/stock/availability_validator.rb (hunk 11 above)
Comment (dkubb) [inline core/app/models/spree/stock/availability_validator.rb]: I should also mention we did previously identify a need for a "quantity" domain within the system, which is simply a natural number, including zero, that would be used for any field that is used for counting. This might be an opportunity to introduce this domain, update all the columns in the system to use it, and then revisit this PR afterwards.

**Comment by dkubb** [comment]
@mbj I finished a quick pass over this and identified a few things that could be extracted into separate commits/PRs, as well as a potential refactoring I'd like to consider before this PR.

File: core/app/models/spree/line_item.rb [hunk 16]
```diff
@@ -52,6 +47,13 @@ def amount
     end
     alias subtotal amount
 
+    # The line item weight
+    #
+    # @return [BigDecimal]
+    def weight
```
Comment (mbj) [inline core/app/models/spree/line_item.rb]: It would have no call side and it would be a really small PR, so semantically it does not add no value to extract it IMO.

File: core/app/models/spree/shipment.rb (hunk 10 above)
Comment (mbj) [inline core/app/models/spree/shipment.rb]: No, its not required. The only call side was removed.

File: core/app/models/spree/stock/availability_validator.rb (hunk 11 above)
Comment (mbj) [inline core/app/models/spree/stock/availability_validator.rb]: I think we should do this in a followup that replaces the stock items with a Boolean on variant.

File: backend/config/reek.yml (hunk 7 above)
Comment (mbj) [inline backend/config/reek.yml]: Its related to the inventory units, so it makes sense to keep it to this commit.

File: core/app/models/spree/app_configuration.rb (hunk 9 above)
Comment (mbj) [inline core/app/models/spree/app_configuration.rb]: Its related to the inventory unit removal IMO, so it to me makes sense to keep it in the same commit.

Comment (mbj) [inline core/app/models/spree/app_configuration.rb]: I could be extracted without breaking specs, but the idea is that when a subsystem is gone all "noun references" get removed also.

File: core/app/models/spree/line_item.rb (hunk 16 above)
Comment (mbj) [inline core/app/models/spree/line_item.rb]: Extracted to #359

File: core/app/models/spree/order.rb (hunk 8 above)
Comment (mbj) [inline core/app/models/spree/order.rb]: Its related to the inventory unit removal, that restructures the way shipment rates are estimated.

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 17]
```diff
@@ -1,106 +1,181 @@
...
-          allow_any_instance_of(ShippingMethod).to receive_message_chain(:calculator, :preferences).and_return({:currency => currency})
-          allow_any_instance_of(ShippingMethod).to receive_message_chain(:calculator, :marked_for_destruction?)
-
-          allow(package).to receive_messages(:shipping_methods => [shipping_method])
-        end
+describe Spree::Stock::Estimator do
+  let(:object)            { described_class.new(stock_location, order)       }
+  let(:calculator_a)      { instance_double(Spree::ShippingCalculator, :a)   }
+  let(:calculator_b)      { instance_double(Spree::ShippingCalculator, :b)   }
+  let(:stock_location)    { instance_double(Spree::StockLocation)            }
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: There's extra spaces before the `{` in these let statements.

**Comment by mbj** [comment]
@dkubb This is ready for review / merge.

File: api/spec/controllers/spree/api/orders_controller_spec.rb [hunk 18]
```diff
@@ -46,7 +46,7 @@ module Spree
         )
       end
 
-      context 'when inventory information is present' do
+      context 'when line_item information is present' do
```
Comment (dkubb) [inline api/spec/controllers/spree/api/orders_controller_spec.rb]: Should this be named "line item" instead? It seems as if the context should be written more from the pov of a human reader firstly.

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 19]
```diff
@@ -1,105 +1,165 @@
...
+      let(:includes_a) { false }
+      let(:includes_b) { false }
+
+      it { should eql([]) }
+    end
+
+    # Case with one shipping method focuses on how a shipping method may produce
+    # a rate and the various fall through behaviors.
+    context 'when address is covered by one shipping method' do
+      let(:includes_a) { true }
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: The ending curly braces are not aligned.

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 20]
```diff
@@ -1,105 +1,165 @@
...
+          )
+          expect(relation).to receive(:new)
+            .with(cost: cost)
+            .and_return(shipping_rate)
         end
 
-        context "when the current matches the order's currency" do
-          it_should_behave_like "shipping rate matches"
+        it 'returns expected rate' do
+          should eql([shipping_rate])
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: Should we be using the implicit `should` or the explicit `is_expected.to` here? Recently I've been favouring the latter.

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 21]
```diff
@@ -1,105 +1,165 @@
...
+        expect(relation).to receive(:new)
+          .with(cost: shipping_rate_b.cost)
+          .and_return(shipping_rate_b)
+      end
 
-            subject.shipping_rates(package)
-          end
-        end
+      it 'returns rates sorted ascending by cost' do
+        should eql([shipping_rate_b, shipping_rate_a])
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66702271

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 22]
```diff
@@ -1,105 +1,165 @@
...
-          let(:currency) { "" }
-          it_should_behave_like "shipping rate matches"
+  describe '.call' do
+    subject { object.call }
+
+    context 'when address is not covered by shipping methods' do
+      let(:includes_a) { false }
+      let(:includes_b) { false }
+
+      it { should eql([]) }
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66702271

File: core/spec/models/spree/stock/estimator_spec.rb [hunk 23]
```diff
@@ -1,105 +1,165 @@
...
+      before do
+        expect(calculator_a).to receive(:available?)
+          .with(package)
+          .and_return(calculator_available)
+      end
+
+      context 'but calculator is not available' do
+        let(:calculator_available) { false }
+
+        it { should eql([]) }
```
Comment (dkubb) [inline core/spec/models/spree/stock/estimator_spec.rb]: Instance of https://github.com/MountainRoseHerbs/spree/pull/329/files#r66702271

**Comment by dkubb** [comment]
@mbj I did a pass over this PR and left a few comments.

File: api/spec/controllers/spree/api/orders_controller_spec.rb (hunk 18 above)
Comment (mbj) [inline api/spec/controllers/spree/api/orders_controller_spec.rb]: yeah, agreed.

**Comment by mbj** [comment]
@dkubb Ready for the next pass
